Implementation of grouped spec files for a single worker instance

RossVertizan opened the pull request:
## Proposed changes
The code is an initial implementation of an ability to group spec files within the "specs' section of the config file using "[ ]", to implement the scheme proposed by @christian-bromann in #6469.  At the time of writing the documentation has not been updated and no unit tests have been created.  However, it does pass tests for mocha, jasmine and cucumber, tested by modifying the examples in the examples/wdio directory.

The main change is to support the array of string arrays structure (string[][]), as well as the existing array of strings (string[]) structure in the definition of the "specs" field in the config file.  This has required changes to the way the specs field is parsed in the configParser (wdio-config) (getFilePaths, getSpecs and setFilePathToFilter methods) to detect that when the items in "specs" are iterated over, that there is now the possibility that some will be arrays.

Also in the launcher (wdio-cli), it has required changes to the way the specs are formatted to be handed over to the workers.  Also a change was necessary in the watcher file (wdio-cli), now creating multiple chokidar.watch jobs (is this OK?).

Since the getSpecs function is shared with other elements e.g. the exclude files and the command line spec field, in places the return type of getSpecs has been asserted to be <string[]>, since we are not proposing supporting this structure in those other places and therefore the return type will always be string[].

First time with TypeScript - I apologise for any clumsiness.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

As per above checklist, no tests or documentation, but I'll add that as the next step.

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)
Solution proposed by @christian-bromann in #6469

### Reviewers: @webdriverio/project-committers

RossVertizan commented:
@christian-bromann as noted in my pull request, I am calling chokidar.watch multiple times and this is causing the tests to fail.  I'm guessing at the functionality of that piece of code, but my guess is that it is watching for changes in the spec files.  So, the alternate solution is to create a single array of all of the spec files, whether they are grouped or not, essentially flatten the hierarchy and uniquify the array.  Does this work as a solution ?
I also need some guidance about adding some tests, do I just add them in the module tests folder?
Am I free to go and make changes to documentation as I see fit?
For both of these are their guidelines?

erwinheitzman commented on packages/wdio-cli/src/launcher.ts:
```diff
@@ -199,7 +199,8 @@ class Launcher {
                 cid: cid++,
                 caps: caps as Capabilities.MultiRemoteCapabilities,
-                specs: this.configParser.getSpecs((caps as Capabilities.DesiredCapabilities).specs, (caps as Capabilities.DesiredCapabilities).exclude).map(s => ({ files: [s], retries: specFileRetries })),
+                // specs: this.configParser.getSpecs((caps as Capabilities.DesiredCapabilities).specs, (caps as Capabilities.DesiredCapabilities).exclude).map(s => ({ files: [s], retries: specFileRetries })),
```
remove commented code

RossVertizan replied:
Yes - will do

erwinheitzman commented on packages/wdio-cli/src/launcher.ts:
```diff
@@ -238,6 +239,24 @@ class Launcher {
+        let returnValue: WorkerSpecs[] = []
+
+        files = this.configParser.getSpecs((capabilities as Capabilities.DesiredCapabilities).specs, (capabilities as Capabilities.DesiredCapabilities).exclude)
+        files.forEach(file => {
```
no need to create a variable and push items, simply map the files and return them

RossVertizan replied:
OK

erwinheitzman commented on packages/wdio-config/src/lib/ConfigParser.ts:
```diff
@@ -316,37 +327,58 @@ export default class ConfigParser {
             patterns = [patterns]
         }
 
+        // Check we have an array
```
No need to state the obvious ;)

christian-bromann replied:
Agree here, either add more context or remove comment completely

erwinheitzman commented on packages/wdio-config/src/lib/ConfigParser.ts:
```diff
@@ -316,37 +327,58 @@ export default class ConfigParser {
+        let groupedFiles: string[] = []
+        let depth: number = hierarchyDepth || 0
 
+        // Convert a string into an array as a convenience
```
No need to state the obvious ;)

RossVertizan replied:
Awww

erwinheitzman commented on packages/wdio-config/src/lib/ConfigParser.ts:
```diff
@@ -316,37 +327,58 @@ export default class ConfigParser {
         }
 
-        patterns = patterns.map(pattern => removeLineNumbers(pattern))
+        // Remove line numbers
```
removeLineNumbers already describes what it does quite well

RossVertizan replied:
Agreed

erwinheitzman commented on packages/wdio-config/src/lib/ConfigParser.ts:
```diff
@@ -316,37 +327,58 @@ export default class ConfigParser {
+            // If pattern is an array, then call getFilePaths again
+            // But only call one level deep, can't have multiple levels of hierarchy
+            if (Array.isArray(pattern) && depth === 0) {
+                // Will always only get a string array back
```
already defined by type

RossVertizan replied:
Ditto

erwinheitzman commented on packages/wdio-config/src/lib/ConfigParser.ts:
```diff
@@ -316,37 +327,58 @@ export default class ConfigParser {
+                groupedFiles = <string[]> ConfigParser.getFilePaths(pattern, omitWarnings, findAndGlob, 1)
+                files.push(groupedFiles)
+            } else {
+                // Otherwise process the files directly
```
please remove this, it doesn't tell the user anything of use

erwinheitzman commented on packages/wdio-config/src/lib/ConfigParser.ts:
```diff
@@ -276,15 +276,26 @@ export default class ConfigParser {
                 fileList.forEach(file => {
-                    if (file.match(filteredFile)) {
-                        filesToFilter.add(file)
+                    if (typeof file === 'string') {
```
files that do not match are silently ignored

RossVertizan replied:
Added a log message

erwinheitzman reviewed (changes requested):
Left some first remarks for you. Is there anything missing from https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md in your opinion? We expect all contributors to read this when contributing to the project  so if there's anything still unclear after reading this, we would like to hear it.

You are free to update anything as it will be reviewed by us before it get's into the codebase :)

RossVertizan commented on packages/wdio-cli/src/launcher.ts:
```diff
@@ -238,6 +239,24 @@ class Launcher {
+    /**
+     * Format the specs into an array of objects with files and retries
+     */
+    formatSpecs(capabilities: (Capabilities.DesiredCapabilities | Capabilities.W3CCapabilities | Capabilities.RemoteCapabilities), specFileRetries: number) {
```
Ah, OK I wasn't clear about the typing of the Capabilities

RossVertizan replied:
Capabilities.Capabilities didn't work, nor Capabilities.RemoteCapabilities

erwinheitzman replied:
let's just keep it as Capabilities.DesiredCapabilities for now then and remove the other two since they aren't used

RossVertizan replied:
When I switch to Capabilities.DesiredCapabilities, I get the following error during transpilation:

```
packages/wdio-cli/src/launcher.ts(202,41): error TS2345: Argument of type 'RemoteCapabilities' is not assignable to parameter of type 'DesiredCapabilities'.
  Type '(DesiredCapabilities | W3CCapabilities)[]' has no properties in common with type 'DesiredCapabilities'.
packages/wdio-cli/src/launcher.ts(214,45): error TS2345: Argument of type 'DesiredCapabilities | W3CCapabilities' is not assignable to parameter of type 'DesiredCapabilities'.
  Type 'W3CCapabilities' has no properties in common with type 'DesiredCapabilities'.
```

christian-bromann replied:
How about:

```suggestion
    formatSpecs(capabilities: Capabilities.RemoteCapability), specFileRetries: number) {
```

RossVertizan replied:
@christian-bromann  - No I get this error:
```
packages/wdio-cli/src/launcher.ts(202,41): error TS2345: Argument of type 'RemoteCapabilities' is not assignable to parameter of type 'RemoteCapability'.
  Type '(W3CCapabilities | DesiredCapabilities)[]' is not assignable to type 'RemoteCapability'.
    Type '(W3CCapabilities | DesiredCapabilities)[]' is not assignable to type 'MultiRemoteCapabilities'.
      Index signature is missing in type '(W3CCapabilities | DesiredCapabilities)[]'.
```

RossVertizan commented on packages/wdio-cli/src/launcher.ts:
```diff
@@ -238,6 +239,24 @@ class Launcher {
+
+        files = this.configParser.getSpecs((capabilities as Capabilities.DesiredCapabilities).specs, (capabilities as Capabilities.DesiredCapabilities).exclude)
+        files.forEach(file => {
+            if (typeof file === 'string') {
```
Sorry, new to Typescript.  Since we are dealing with an array of strings or an array of string arrays are we not guaranteed to always be getting either a string or an array? I don't know what runtime type checking there is on user data.  I would normally always implement the drop through to catch anything else, but omitted it here as overkill.  In the next push, I have implemented such a drop through.

RossVertizan replied:
For type checking I now have to return an empty entry, it would be good we could drop anything that is not a string or an array - please see the new code when I push it.

RossVertizan replied:
@erwinheitzman Can we close out this conversation ?

RossVertizan commented on packages/wdio-cli/src/watcher.ts:
```diff
@@ -12,7 +12,7 @@ const log = logger('@wdio/cli:watch')
 export default class Watcher {
     private _launcher: Launcher
-    private _specs: string[]
+    private _specs: ( string | string[])[]
```
Sure.  Where do I declare types.  I have this code, but don't know where it goes:
```
/** Define the types to be used in the specs array
 *  string[] - array of strings
 *  string[][] - array of string arrays
 */
export declare type SpecsArrayEntry =
    | string
    | string[];

export declare type SpecsArray =
    SpecsArrayEntry[];
```

erwinheitzman replied:
Above the method:
```
type Spec = string | string[]
```

```suggestion
    private _specs: Spec[]
```

RossVertizan commented on packages/wdio-config/src/lib/ConfigParser.ts:
```diff
@@ -226,7 +226,8 @@ export default class ConfigParser {
                 }
                 if (Array.isArray(suite)) {
-                    suiteSpecs = suiteSpecs.concat(ConfigParser.getFilePaths(suite, undefined, this._pathService))
+                    // Not supporting hierarchical suites - return will always be string[]
```
The return type from getFilePaths is (string | string[])[], which can't be assigned to suiteSpecs which is type string[].  However, since we are not proposing to introduce the same kind of hierarchy into the suite spec, we know the return value will always be string[], and so by asserting it, this will compile.  As I said before, I am new to TypeScript, so there may be a better way to approach this, please advise.

RossVertizan replied:
@erwinheitzman Can we close out this conversation

RossVertizan commented on packages/wdio-config/src/lib/ConfigParser.ts:
```diff
@@ -276,15 +276,26 @@ export default class ConfigParser {
             filteredFile = removeLineNumbers(filteredFile)
-            let globMatchedFiles = ConfigParser.getFilePaths(this._pathService.glob(filteredFile), undefined, this._pathService)
+            // Send single file/file glob to getFilePaths - not supporting hierarchy in spec
+            // Return value will alwyas be string[]
```
Ditto above comment

RossVertizan commented:
After the latest round of changes, I'm getting this error:
`packages/webdriverio/src/utils/findStrategy.ts(3,111): error TS2307: Cannot find module 'aria-query' or its corresponding type declarations.`
I don't know what has caused me to start getting this error, any hints?

RossVertizan commented:
For the watcher, I intend to flatten the array for the chokidar.watch, then in the callback traversing the this._specs to find if the file is a single file or part of an array (or both) and send the necessary calls to "run".  In the current specs scheme, if there are duplicated files, then a single call to run the file is all that is ever required.  In the case of this new scheme, we may have the file as a single file, but also part of an array and so we will need to make multiple "run" calls.

RossVertizan commented:
>Is there anything missing from https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md in your opinion?

Where should I make those comments?

christian-bromann commented:
> Where should I make those comments?

Maybe create a separate issue?

erwinheitzman commented:
@RossVertizan is there anything you need in order to have this PR move forward?

RossVertizan commented:
@erwinheitzman Thanks for reaching out.  I think I am done now with at least something in the smoke tests and some jest tests for wdio-config, but see my issue #6593 for the wdio-cli tests - I have one commented out at the moment (formatSpecs).  What is the process for resolving your comments?  I think I have responded to everything, either by making a code change or by explaining myself.  So, I kind of need guidance from you guys as to how to proceed.

christian-bromann commented on packages/wdio-cli/tests/launcher.test.ts:
```diff
@@ -269,6 +269,17 @@ describe('launcher', () => {
+    //         // Get: [{"files": ["./tests/test1.js"], "retries": 17}]
+    //         expect(launcher.formatSpecs(capabilities, specFileRetries)).toStrictEqual(expected)
+    //     })
+    // })
```
What happens with this unit test?

RossVertizan replied:
1. Using Capabilities.RemoteCapability results in this transpile error:
```
packages/wdio-cli/src/launcher.ts(202,41): error TS2345: Argument of type 'RemoteCapabilities' is not assignable to parameter of type 'DesiredCapabilities | W3CCapabilities | MultiRemoteCapabilities'.
  Type '(DesiredCapabilities | W3CCapabilities)[]' is not assignable to type 'DesiredCapabilities | W3CCapabilities | MultiRemoteCapabilities'.
    Type '(DesiredCapabilities | W3CCapabilities)[]' is not assignable to type 'MultiRemoteCapabilities'.
      Index signature is missing in type '(DesiredCapabilities | W3CCapabilities)[]'.
```

2. Stating the obvious, the unit test does nothing, as it is commented.  Being a little more helpful, given the input I would expect this output:
```[{"files": ["/a.js"], "retries": 17}, {"files": ["/b.js", "/c.js"], "retries": 17}]```
but in actual fact get this:
```[{"files": ["./tests/test1.js"], "retries": 17}]```
this is the same issue that I filed in this ticket #6593

christian-bromann replied:
I commented in the issue. Most of the classes are mocked and don't always reflect original behavior.

christian-bromann commented on packages/wdio-cli/tests/__fixtures__/array.conf.js:
```diff
@@ -0,0 +1,20 @@
+import path from 'path'
```
I don't see this file being used anywhere

RossVertizan replied:
Again #6593.  It doesn't seem to matter what's in the config file I always get the same result, so I stopped developing this line of testing.

RossVertizan replied:
I have now deleted this file as the contents are not being used.  Rather the return value of getSpecs is being mocked.

christian-bromann commented on tests/smoke.runner.js:
```diff
@@ -12,7 +12,7 @@ const fs = {
 }
 
-import launch from './helpers/launch'
+import  launch from './helpers/launch'
```
```suggestion
import launch from './helpers/launch'
```

christian-bromann commented on tests/smoke.runner.js:
```diff
@@ -347,6 +347,45 @@ const jasmineSpecFiltering = async () => {
+    assert.strictEqual(skippedSpecs, 0)
+}
+
+/**
+ * Jasmine with spec grouping feature enabled
+ */
+const jasmineSpecGrouping = async () => {
+    const { skippedSpecs } = await launch(
+        path.resolve(__dirname, 'helpers', 'config.js'),
+        {
+            specs: [
+                [
+                    path.resolve(__dirname, 'jasmine', 'test.js'),
+                    path.resolve(__dirname, 'jasmine', 'test-skipped.js'),
+                    path.resolve(__dirname, 'jasmine', 'test-skipped-grep.js')
+                ]
+            ],
+            framework: 'jasmine'
+        })
+    // Specs will be treated as a group, so no specs will be skipped
+    assert.strictEqual(skippedSpecs, 0)
+}
```
This feature is independent from which framework is being run, so we can remove one smoke test
```suggestion
```

RossVertizan replied:
Yes agreed

christian-bromann commented on tests/smoke.runner.js:
```diff
@@ -347,6 +347,45 @@ const jasmineSpecFiltering = async () => {
+                ]
+            ]
+        })
+    // Specs will be treated as a group, so no specs will be skipped
```
Why are specs in group skipped?

RossVertizan replied:
Specs in group will **not** be skipped.
```assert.strictEqual(skippedSpecs, 0)```

christian-bromann replied:
Why is that? If I have my config defined as follows:

```js
specs: [
  "/specA.e2e.js",
  ["/specA.e2e.js"]
]
```

where `specA.e2e.js` has this content:

```js
describe.skip("my feature", () => {
  // ...
})
```

It would skip tests for both files. I think I am missing something here.

RossVertizan replied:
No, that is right.  In that specific case because "all" the specs in the group would be skipped, then the group would be skipped (no specs to run).  However, in the specific case of the test case, there are specs in the group that will be run and so the group spec will be run and so the returned value of specs skipped will be 0.

Is this desired behaviour? it seems correct to me.

Maybe worth another test case?

christian-bromann replied:
No you are right, I understand now 👍

christian-bromann commented:
@RossVertizan I will take another look at the PR and the other issue tomorrow. Thanks for following up on the comments. One question: given this is a high demanded feature (I can't count how many folks have asked how to run tests sequentially) would you mind to make a 2nd PR adding a small blog post announcing/showing this feature and explaining why it is so useful?

RossVertizan commented:
Sure

christian-bromann commented:
@RossVertizan thanks so much. I think this will help folks to understand what this feature is all about and why it is so useful. Please add a markdown file with the content of your post here: https://github.com/webdriverio/webdriverio/tree/main/website/blog. Again it doesn't need to be a long post, just some text highlighting that if you want to run specs in a sequential manner you can group them now.

I added a PR to update the spec reporter making use of multiple spec files: https://github.com/RossVertizan/webdriverio/pull/1

christian-bromann commented:
Lastly, it would be great if we could have a section on grouping specs in https://webdriver.io/docs/organizingsuites

RossVertizan commented:
I the latest push I have:
- added tests in Jest for all of the new features
- added the blog post - ready for review
- added a section to the Organizing Test Suites document

Can I execute the PR request you filed?

All in all I think this is now ready for another review

christian-bromann commented:
> Can I execute the PR request you filed?

Yes, let's get that one merged into the branch.

> All in all I think this is now ready for another review

Will take another look, thank you a lot!

nextlevelbeard commented:
Does this play well in conjunction with the specs per capability feature? Also the exclude option?

RossVertizan commented:
@nextlevelbeard thanks good catch.  Yes, the code already handled the specs per capability functionality.  However, I had overlooked the exclude functionality.  Added in the latest push.

RossVertizan commented:
Now seeing the CI checks fail with these messages.  I've not touched anything in the Jasmine framework.
```
Error: ../../../packages/wdio-jasmine-framework/node_modules/@wdio/types/build/index.d.ts(23,13): error TS2374: Duplicate string index signature.
Error: ../../../packages/wdio-jasmine-framework/node_modules/@wdio/types/build/index.d.ts(26,13): error TS2374: Duplicate string index signature.
Error: ../../../packages/wdio-jasmine-framework/node_modules/@wdio/types/build/index.d.ts(29,13): error TS2374: Duplicate string index signature.
Error: ../../../packages/wdio-jasmine-framework/node_modules/@wdio/types/build/index.d.ts(43,14): error TS2300: Duplicate identifier 'Capabilities'.
Error: ../../../packages/wdio-jasmine-framework/node_modules/@wdio/types/build/index.d.ts(44,14): error TS2300: Duplicate identifier 'DesiredCapabilities'.
Error: ../../../packages/wdio-types/build/index.d.ts(35,19): error TS2320: Interface 'Config' cannot simultaneously extend types 'Testrunner' and 'Testrunner'.
  Named property 'after' of types 'Testrunner' and 'Testrunner' are not identical.
Error: ../../../packages/wdio-types/build/index.d.ts(35,19): error TS2320: Interface 'Config' cannot simultaneously extend types 'Testrunner' and 'Testrunner'.
  Named property 'afterSession' of types 'Testrunner' and 'Testrunner' are not identical.
Error: ../../../packages/wdio-types/build/index.d.ts(35,19): error TS2320: Interface 'Config' cannot simultaneously extend types 'Testrunner' and 'Testrunner'.
  Named property 'before' of types 'Testrunner' and 'Testrunner' are not identical.
Error: ../../../packages/wdio-types/build/index.d.ts(35,19): error TS2320: Interface 'Config' cannot simultaneously extend types 'Testrunner' and 'Testrunner'.
  Named property 'beforeSession' of types 'Testrunner' and 'Testrunner' are not identical.
Error: ../../../packages/wdio-types/build/index.d.ts(35,19): error TS2320: Interface 'Config' cannot simultaneously extend types 'Testrunner' and 'Testrunner'.
  Named property 'capabilities' of types 'Testrunner' and 'Testrunner' are not identical.
...
```
There must be something else changing that is causing these failures?

christian-bromann commented:
@RossVertizan can you try to update your branch?

RossVertizan commented:
@christian-bromann my branch is already up to date with webdriverio:main

christian-bromann commented on examples/wdio/jasmine/wdio.conf.js:
```diff
@@ -1,10 +1,15 @@
 const path = require('path')
 
 exports.config = {
+
+    /**
+     * server configurations
+     */
+
```
This doesn't seem to be needed.
```suggestion
```

christian-bromann commented on foo.txt:
```diff
@@ -0,0 +1,125 @@
+
```
Let's remove this file, it comes from my PR 🙈

christian-bromann commented on package.json:
```diff
@@ -101,5 +101,7 @@
     "url-join": "^4.0.1"
   },
-  "dependencies": {}
+  "dependencies": {
+    "@types/aria-query": "^4.2.1"
+  }
```
This seems unrelated to your PR, no?

RossVertizan replied:
Indeed, though as noted previously I sometimes see transpilation fail when I don't have the aria types installed

christian-bromann commented on packages/wdio-cli/src/watcher.ts:
```diff
@@ -74,10 +76,31 @@ export default class Watcher {
      */
     getFileListener (passOnFile = true) {
         return (spec: string) => {
+            let runSpecs: Spec[] = []
```
```suggestion
            const runSpecs: Spec[] = []
```

christian-bromann commented on packages/wdio-config/src/lib/ConfigParser.ts:
```diff
@@ -314,39 +326,75 @@ export default class ConfigParser {
-                    (ext) => filename.endsWith(ext)))
+            // If pattern is an array, then call getFilePaths again
+            // But only call one level deep, can't have multiple levels of hierarchy
+            if (Array.isArray(pattern) && depth === 0) {
```
Can't we just do 
```suggestion
            if (Array.isArray(pattern) && hierarchyDepth) {
```
and get rid of `depth`?

RossVertizan replied:
Agreed

christian-bromann commented on packages/wdio-jasmine-framework/package.json:
```diff
@@ -27,8 +27,9 @@
     "@wdio/logger": "7.0.0",
-    "@wdio/types": "7.2.1",
-    "@wdio/utils": "7.2.1",
+    "@wdio/types": "7.2.0",
+    "@wdio/utils": "7.2.0",
+    "callsites": "^3.1.0",
```
```suggestion
```

christian-bromann commented on packages/wdio-jasmine-framework/src/index.ts:
```diff
@@ -102,6 +102,19 @@ class JasmineAdapter {
          * enable expectHandler
          */
         jasmine.Spec.prototype.addExpectationResult = this.getExpectationResultHandler(jasmine)
+        Object.defineProperty(global, '__filename', {
+            get: function () {
+                // @ts-expect-error
+                let path = __stack[1].getFileName()
+                console.log('YPP', path)
+
+                try { //*nix OSes
+                    return path.match(/[^/]+\/[^/]+$/)[0]
+                } catch (error) { //Windows based OSes
+                    return path.match(/[^\\]+\\[^\\]+$/)[0]
+                }
+            }
+        })
```
```suggestion
```

christian-bromann commented on packages/wdio-jasmine-framework/src/index.ts:
```diff
@@ -251,6 +264,11 @@ class JasmineAdapter {
             this._jrunner.env.beforeAll(this.wrapHook('beforeSuite'))
             this._jrunner.env.afterAll(this.wrapHook('afterSuite'))
 
+            console.log(this._jrunner.env)
+
+            // const runner = this._jrunner.env.currentRunner
+            // console.log(111, runner.specs())
+
```
```suggestion
```

christian-bromann commented on packages/wdio-jasmine-framework/src/reporter.ts:
```diff
@@ -45,6 +45,8 @@ export default class JasmineReporter {
     }
 
     specStarted (test: jasmine.CustomReporterResult) {
+        // console.log(test);
+
```
```suggestion
```

christian-bromann commented on packages/wdio-reporter/src/index.ts:
```diff
@@ -49,6 +49,7 @@ export default class WDIOReporter extends EventEmitter {
         const rootSuite = new SuiteStats({
             title: '(root)',
             fullTitle: '(root)',
+            file: ''
```
```suggestion
```

christian-bromann commented on packages/wdio-reporter/src/stats/suite.ts:
```diff
@@ -26,6 +26,7 @@ export default class SuiteStats extends RunnableStats {
     cid?: string
     title: string
     fullTitle: string
+    file: string
```
```suggestion
```

christian-bromann commented on packages/wdio-reporter/src/stats/suite.ts:
```diff
@@ -43,6 +44,7 @@ export default class SuiteStats extends RunnableStats {
         this.title = suite.title
         this.fullTitle = suite.fullTitle
         this.tags = suite.tags
+        this.file = suite.file
```
```suggestion
```

christian-bromann commented on website/docs/OrganizingTestSuites.md:
````diff
@@ -102,6 +102,26 @@ Or, run multiple suites at once:
+        './test/specs/test_b*.js',
+    ],
+```
+In the example above, all of the "test_a*" tests will be run sequentially in a single
+instance and each of the "test_b*" tests will run concurrently in individual instances.
````
```suggestion
In the example above, all of the "test_a*" tests will be run sequentially in a single instance and each of the "test_b*" tests will run concurrently in individual instances.
```

christian-bromann commented on website/docs/OrganizingTestSuites.md:
```diff
@@ -102,6 +102,26 @@ Or, run multiple suites at once:
+transpiling code or provisioning cloud instances, but there are also advanced usage
+models that benefit from this capability.
+
+To group tests to run in a single instance, simply define them as an array within
+the specs definition.
```
```suggestion
To group tests to run in a single instance, simply define them as an array within the specs definition.
```

christian-bromann reviewed (commented):
some more formatting corrections.

RossVertizan commented:
All changes made

cpeck-bi commented:
From the failing tests-- are we missing dependencies for the reporters? I don't see them in package.json

christian-bromann commented on packages/wdio-jasmine-framework/package.json:
```diff
@@ -27,8 +27,6 @@
   "dependencies": {
     "@types/jasmine": "3.6.7",
     "@wdio/logger": "7.0.0",
-    "@wdio/types": "7.2.1",
-    "@wdio/utils": "7.2.1",
```
This change should be reverted

christian-bromann commented on packages/wdio-jasmine-framework/src/reporter.ts:
```diff
@@ -45,6 +45,7 @@ export default class JasmineReporter {
     }
 
     specStarted (test: jasmine.CustomReporterResult) {
+
```
```suggestion
```

christian-bromann commented on packages/wdio-jasmine-framework/package.json:
```diff
@@ -27,8 +27,6 @@
   "dependencies": {
     "@types/jasmine": "3.6.7",
     "@wdio/logger": "7.0.0",
```
```suggestion
    "@wdio/logger": "7.0.0",
    "@wdio/types": "7.2.1",
    "@wdio/utils": "7.2.1",
```

christian-bromann commented on website/blog/2021-03-23-grouping-specs.md:
````diff
@@ -0,0 +1,57 @@
+So, the following specs definition:
+
+```json
+    "specs": [
+                [
+                    './test/specs/test_login.js',
+                    './test/specs/test_product_order.js',
+                    './test/specs/test_checkout.js'
+                ],
+                './test/specs/test_b*.js',
+            ],
````
```suggestion
    "specs": [
        [
            "./test/specs/test_login.js",
            "./test/specs/test_product_order.js",
            "./test/specs/test_checkout.js"
        ],
        "./test/specs/test_b*.js",
    ],
```

christian-bromann commented on website/docs/OrganizingTestSuites.md:
````diff
@@ -102,6 +102,24 @@ Or, run multiple suites at once:
+To group tests to run in a single instance, simply define them as an array within the specs definition.
+
+```json
+  "specs": [
+                [
+                    './test/specs/test_login.js',
+                    './test/specs/test_product_order.js',
+                    './test/specs/test_checkout.js'
+                ],
+                './test/specs/test_b*.js',
+            ],
````
```suggestion
    "specs": [
        [
            "./test/specs/test_login.js",
            "./test/specs/test_product_order.js",
            "./test/specs/test_checkout.js"
        ],
        "./test/specs/test_b*.js",
    ],
```

RossVertizan commented:
@christian-bromann what do I need to do to get this PR closed?  I'd like to get this wrapped up, if possible.

christian-bromann commented:
@RossVertizan there seems to be a typing issue:

```
Error: packages/wdio-reporter/src/index.ts(49,42): error TS2345: Argument of type '{ title: string; fullTitle: string; }' is not assignable to parameter of type 'Suite'.
  Property 'file' is missing in type '{ title: string; fullTitle: string; }' but required in type 'Suite
```

this needs to get fixed first. I can take a look later today.

christian-bromann commented on packages/wdio-reporter/src/stats/suite.ts:
```diff
@@ -9,7 +9,7 @@ export interface Suite {
     fullTitle: string
     pending?: boolean
-    fil?: string
+    file: string
```
```suggestion
    file?: string
```

christian-bromann reviewed (commented):
fix the type bug

RossVertizan commented:
@christian-bromann please let me know if there is anything I can do

christian-bromann commented:
@RossVertizan I just created a PR for your fork to fix the errors

RossVertizan commented:
@christian-bromann I merged your PR, but now we get different failures.  Am I doing something wrong?

christian-bromann commented on packages/wdio-reporter/src/stats/suite.ts:
```diff
@@ -9,7 +9,7 @@ export interface Suite {
     fullTitle: string
     pending?: boolean
-    fil?: string
+    file?: string
```
```suggestion
    file: string
```

christian-bromann commented:
@RossVertizan I think the failing tests are legit and were introduce by making the recent adjustments. The problematic part is:

```ts
        for (let pattern of patterns) {
            // If pattern is an array, then call getFilePaths again
            // But only call one level deep, can't have multiple levels of hierarchy
            if (Array.isArray(pattern) && hierarchyDepth) {
                // Will always only get a string array back
                groupedFiles = <string[]>ConfigParser.getFilePaths(pattern, omitWarnings, findAndGlob, 1)
                files.push(groupedFiles)
            } else {
                let filenames = findAndGlob.glob(<string>pattern)
                // ...
```

In the last line (`let filenames = findAndGlob.glob(<string>pattern)`, `pattern` is an array of strings which will fails in the glob module as it only accepts a string. Something needs to be adjusted here with the `hierarchyDepth` handling.

RossVertizan commented:
@christian-bromann  - OK, I'll take a look into it.

RossVertizan commented:
@christian-bromann - I believe I have solved the problem in the getFilePaths recursion.  It was actually a coding error.

I now have a problem with the ```packages/wdio-reporter/tests/stats/suite.test.ts```.

```jest
 FAIL  packages/wdio-reporter/tests/stats/suite.test.ts
  ● should get initialised

    expect(received).toMatchSnapshot()

    Snapshot name: `should get initialised 1`

    - Snapshot  - 0
    + Received  + 1

    @@ -1,9 +1,10 @@
      Object {
        "_duration": 0,
        "cid": "0-0",
        "description": "some description",
    +   "file": undefined,
        "fullTitle": "barfoo",
        "hooks": Array [],
        "hooksAndTests": Array [],
        "suites": Array [],
        "tags": Array [
```
The received value has the file line.  Is this correct?  Should I update the snapshot?  Can I manually edit the snapshot file?  Is the received value correct?

christian-bromann commented:
> The received value has the file line. Is this correct? Should I update the snapshot? Can I manually edit the snapshot file? Is the received value correct?

Can you add a file property to the fixture so that the file property shows some content?

RossVertizan commented:
I'll give it a go.

christian-bromann commented:
Meaning, adding a file property here: https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-reporter/tests/stats/suite.test.ts#L4-L10, then run the test with `-u` to update the snapshot:

```sh
npx jest packages/wdio-reporter/tests/stats/suite.test.ts -u --collectCoverage=false
```

RossVertizan commented:
Yes - doing it right now.  Thanks for the -u advice.

RossVertizan commented:
@christian-bromann: It's now failing the smoke tests because suite.file is undefined in ```wdio-spec-reporter/build/index.js``` [here](https://github.com/RossVertizan/webdriverio/blob/525faeafe0bf347f8b911ed75458f9deb1c69095/packages/wdio-spec-reporter/src/index.ts#L216-L220)

```
// Display file path of spec
if (!specFileReferences.includes(suite.file)) {
    output.push(`${suiteIndent}» ${suite.file.replace(process.cwd(), '')}`)
    specFileReferences.push(suite.file)
}
```

Add a check for it being undefined?  Or define it in the test?  Where?

christian-bromann commented:
@RossVertizan added another PR that hopefully will fix it: https://github.com/RossVertizan/webdriverio/pull/3/files

christian-bromann commented on website/blog/2021-03-23-grouping-specs.md:
```diff
@@ -0,0 +1,56 @@
+
+However, there are cases where this does not work so well.  In one case a users flow involved transpiling tens of thousands of Typescript files for each of the~250 test files, resulting in a huge overhead in the speed of the testing.  In another case a remote device farm was provisioning a new device for each test with all of the associated setup, impacting performance and cost.
+
+At Vertizan (https://www.vitaq.io) we are integrating our AI-driven and coverage-led Vitaq test automation tool with WebdriverIO and Mocha. For Vitaq AI to work, it needs to be able to select which test/action to run next and that requires having all of the tests available in a single instance.
```
```suggestion
At [Vertizan](https://www.vitaq.io) we are integrating our AI-driven and coverage-led Vitaq test automation tool with WebdriverIO and Mocha. For Vitaq AI to work, it needs to be able to select which test/action to run next and that requires having all of the tests available in a single instance.
```

christian-bromann reviewed (approved):
This was a big effort! Thank you so much @RossVertizan for tackling it so excellent 👏. I would love to always get such great first time contributions. Hopefully there are more to come!

RossVertizan commented:
@christian-bromann Thanks for all your help, I have learnt so much.  I do feel as though you could have done this in about half an hour, but I guess that's not the point, at least I am up and running, well, standing anyway.

I have just made some final tweaks to the docs, can I raise another PR just to get them in? In particular there was an error in the "Organizing Test Suites".

Now that this is done, I think I'll circle back around to the beginning and try to understand why the smoke tests were failing on my machine.  I'm guessing you are using a *nix flavour?

RossVertizan commented:
Created a new PR #6638
